hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java

```diff
@@ -89,4 +119,30 @@ private Message process(final SCMRatisRequest request)
     }
   }
 
+  @Override
+  public long takeSnapshot() throws IOException {
+    long startTime = Time.monotonicNow();
+    TermIndex lastTermIndex = getLastAppliedTermIndex();
+    long lastAppliedIndex = lastTermIndex.getIndex();
+    SCMTransactionInfo lastAppliedTrxInfo =
+        SCMTransactionInfo.fromTermIndex(lastTermIndex);
+    if (transactionBuffer.getLatestTrxInfo()
+        .compareTo(lastAppliedTrxInfo) < 0) {
+      transactionBuffer.updateLatestTrxInfo(
+          SCMTransactionInfo.builder()
+              .setCurrentTerm(lastTermIndex.getTerm())
+              .setTransactionIndex(lastTermIndex.getIndex())
+              .build());
+      transactionBuffer.setLatestSnapshot(
+          transactionBuffer.getLatestTrxInfo().toSnapshotInfo());
+    } else {
+      lastAppliedIndex =
+          transactionBuffer.getLatestTrxInfo().getTransactionIndex();
+    }
+
+    transactionBuffer.flush();
+    LOG.info("Current Snapshot Index {}, takeSnapshot took {} ms",
+        getLastAppliedTermIndex(), Time.monotonicNow() - startTime);
+    return lastAppliedIndex;
```

Comment:
why the lastApplied index returned here is the one derived from transaction buffer? Once the snapshot is taken, the last applied index should be given by getLastAppliedTermIndex() not transaction buffer which may be behind actual last applied index in ratis?

Reply:
O the code here is keeping the lastAppliedIndex in buffer to equal to the getLastAppliedTermIndex() actually. 

```
    LOG.info("Current Snapshot Index {}, takeSnapshot took {} ms",
        getLastAppliedTermIndex(), Time.monotonicNow() - startTime);
```
is wrong.

I should replace `getLastAppliedTermIndex()` with `lastAppliedIndex` to make it clear.


We need the `lastAppliedIndex` in buffer keep updated so that will be written into DB.

Reply:
I see the buffer getting updated with applied index only after applyTransaction is called. ApplyTransaction is not invoked for every log transaction in Ratis. So, its better to rely on getLastAppliedTermIndex() rather than the buffer for the latest applied index.